### ISSUE TITLE
Problem Upgrading to 0.3.8

### ISSUE THREAD

**Issue by cheeseng** [scala-native/scala-native#1400]
Hi All, 

ScalaTest team is using 0.3.6 and trying to update to the latest 0.3.8 but unfortunately we encountered some problem.

In 0.3.6, we have the following code in our build to skip the 2 optimization passes: 

https://github.com/cheeseng/scalatest/blob/experiment-scala-native-0.3.8/project/scalatest.scala#L679

which solved a runtime crash when we're executing our tests, which the crash looks like the following: 

```
java.net.SocketException: Socket is closed
	at java.net.Socket.setSoTimeout(Socket.java:1137)
	at scala.scalanative.testinterface.ComRunner.receive(ComRunner.scala:101)
	at scala.scalanative.testinterface.ScalaNativeRunner.receive(ScalaNativeRunner.scala:36)
	at scala.scalanative.testinterface.ScalaNativeTask.receive$1(ScalaNativeTask.scala:31)
	at scala.scalanative.testinterface.ScalaNativeTask.execute(ScalaNativeTask.scala:42)
	at sbt.TestRunner.runTest$1(TestFramework.scala:76)
	at sbt.TestRunner.run(TestFramework.scala:85)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1$$anonfun$apply$8.apply(TestFramework.scala:202)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1$$anonfun$apply$8.apply(TestFramework.scala:202)
	at sbt.TestFramework$.sbt$TestFramework$$withContextLoader(TestFramework.scala:185)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1.apply(TestFramework.scala:202)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1.apply(TestFramework.scala:202)
	at sbt.TestFunction.apply(TestFramework.scala:207)
	at sbt.Tests$.sbt$Tests$$processRunnable$1(Tests.scala:239)
	at sbt.Tests$$anonfun$makeSerial$1.apply(Tests.scala:245)
	at sbt.Tests$$anonfun$makeSerial$1.apply(Tests.scala:245)
	at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
	at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
[error] Could not run test org.scalactic.DifferSpec: java.net.SocketException: Socket is closed
java.lang.NullPointerException
	at sbt.JUnitXmlTestsListener$TestSuite$$anonfun$2.apply(JUnitXmlTestsListener.scala:78)
	at sbt.JUnitXmlTestsListener$TestSuite$$anonfun$2.apply(JUnitXmlTestsListener.scala:71)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
	at scala.collection.immutable.List.foreach(List.scala:318)
	at scala.collection.generic.TraversableForwarder$class.foreach(TraversableForwarder.scala:32)
	at scala.collection.mutable.ListBuffer.foreach(ListBuffer.scala:45)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:244)
	at scala.collection.AbstractTraversable.map(Traversable.scala:105)
	at sbt.JUnitXmlTestsListener$TestSuite.stop(JUnitXmlTestsListener.scala:71)
	at sbt.JUnitXmlTestsListener.writeSuite(JUnitXmlTestsListener.scala:180)
	at sbt.JUnitXmlTestsListener.endGroup(JUnitXmlTestsListener.scala:161)
	at sbt.TestRunner$$anonfun$run$4.apply(TestFramework.scala:90)
	at sbt.TestRunner$$anonfun$run$4.apply(TestFramework.scala:90)
	at sbt.TestFramework$$anonfun$safeForeach$1.apply(TestFramework.scala:107)
	at sbt.TestFramework$$anonfun$safeForeach$1.apply(TestFramework.scala:107)
	at scala.collection.immutable.List.foreach(List.scala:318)
	at sbt.TestFramework$.safeForeach(TestFramework.scala:107)
	at sbt.TestRunner.safeListenersCall(TestFramework.scala:96)
	at sbt.TestRunner.run(TestFramework.scala:90)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1$$anonfun$apply$8.apply(TestFramework.scala:202)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1$$anonfun$apply$8.apply(TestFramework.scala:202)
	at sbt.TestFramework$.sbt$TestFramework$$withContextLoader(TestFramework.scala:185)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1.apply(TestFramework.scala:202)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1.apply(TestFramework.scala:202)
	at sbt.TestFunction.apply(TestFramework.scala:207)
	at sbt.Tests$.sbt$Tests$$processRunnable$1(Tests.scala:239)
	at sbt.Tests$$anonfun$makeSerial$1.apply(Tests.scala:245)
	at sbt.Tests$$anonfun$makeSerial$1.apply(Tests.scala:245)
	at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
	at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
[error] java.lang.NullPointerException
java.net.SocketException: Socket closed
	at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:118)
	at java.net.SocketOutputStream.write(SocketOutputStream.java:155)
	at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
	at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
	at java.io.DataOutputStream.flush(DataOutputStream.java:123)
	at java.io.FilterOutputStream.close(FilterOutputStream.java:158)
	at scala.scalanative.testinterface.ComRunner.close(ComRunner.scala:107)
	at scala.scalanative.testinterface.ComRunner.send(ComRunner.scala:63)
	at scala.scalanative.testinterface.ScalaNativeRunner.send(ScalaNativeRunner.scala:33)
	at scala.scalanative.testinterface.ScalaNativeTask.execute(ScalaNativeTask.scala:27)
	at sbt.TestRunner.runTest$1(TestFramework.scala:76)
	at sbt.TestRunner.run(TestFramework.scala:85)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1$$anonfun$apply$8.apply(TestFramework.scala:202)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1$$anonfun$apply$8.apply(TestFramework.scala:202)
	at sbt.TestFramework$.sbt$TestFramework$$withContextLoader(TestFramework.scala:185)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1.apply(TestFramework.scala:202)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1.apply(TestFramework.scala:202)
	at sbt.TestFunction.apply(TestFramework.scala:207)
	at sbt.Tests$.sbt$Tests$$processRunnable$1(Tests.scala:239)
	at sbt.Tests$$anonfun$makeSerial$1.apply(Tests.scala:245)
	at sbt.Tests$$anonfun$makeSerial$1.apply(Tests.scala:245)
	at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
	at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
[error] Could not run test org.scalactic.anyvals.NonEmptySetSpec: java.net.SocketException: Socket closed
java.lang.NullPointerException
	at sbt.JUnitXmlTestsListener$TestSuite$$anonfun$2.apply(JUnitXmlTestsListener.scala:78)
	at sbt.JUnitXmlTestsListener$TestSuite$$anonfun$2.apply(JUnitXmlTestsListener.scala:71)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
	at scala.collection.immutable.List.foreach(List.scala:318)
	at scala.collection.generic.TraversableForwarder$class.foreach(TraversableForwarder.scala:32)
	at scala.collection.mutable.ListBuffer.foreach(ListBuffer.scala:45)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:244)
	at scala.collection.AbstractTraversable.map(Traversable.scala:105)
	at sbt.JUnitXmlTestsListener$TestSuite.stop(JUnitXmlTestsListener.scala:71)
	at sbt.JUnitXmlTestsListener.writeSuite(JUnitXmlTestsListener.scala:180)
	at sbt.JUnitXmlTestsListener.endGroup(JUnitXmlTestsListener.scala:161)
	at sbt.TestRunner$$anonfun$run$4.apply(TestFramework.scala:90)
	at sbt.TestRunner$$anonfun$run$4.apply(TestFramework.scala:90)
	at sbt.TestFramework$$anonfun$safeForeach$1.apply(TestFramework.scala:107)
	at sbt.TestFramework$$anonfun$safeForeach$1.apply(TestFramework.scala:107)
	at scala.collection.immutable.List.foreach(List.scala:318)
	at sbt.TestFramework$.safeForeach(TestFramework.scala:107)
	at sbt.TestRunner.safeListenersCall(TestFramework.scala:96)
	at sbt.TestRunner.run(TestFramework.scala:90)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1$$anonfun$apply$8.apply(TestFramework.scala:202)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1$$anonfun$apply$8.apply(TestFramework.scala:202)
	at sbt.TestFramework$.sbt$TestFramework$$withContextLoader(TestFramework.scala:185)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1.apply(TestFramework.scala:202)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1.apply(TestFramework.scala:202)
	at sbt.TestFunction.apply(TestFramework.scala:207)
	at sbt.Tests$.sbt$Tests$$processRunnable$1(Tests.scala:239)
	at sbt.Tests$$anonfun$makeSerial$1.apply(Tests.scala:245)
	at sbt.Tests$$anonfun$makeSerial$1.apply(Tests.scala:245)
	at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
	at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
[error] java.lang.NullPointerException
java.net.SocketException: Socket closed
	at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:118)
	at java.net.SocketOutputStream.write(SocketOutputStream.java:155)
	at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
	at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
	at java.io.DataOutputStream.flush(DataOutputStream.java:123)
	at java.io.FilterOutputStream.close(FilterOutputStream.java:158)
	at scala.scalanative.testinterface.ComRunner.close(ComRunner.scala:107)
	at scala.scalanative.testinterface.ComRunner.send(ComRunner.scala:63)
	at scala.scalanative.testinterface.ScalaNativeRunner.send(ScalaNativeRunner.scala:33)
	at scala.scalanative.testinterface.ScalaNativeTask.execute(ScalaNativeTask.scala:27)
	at sbt.TestRunner.runTest$1(TestFramework.scala:76)
	at sbt.TestRunner.run(TestFramework.scala:85)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1$$anonfun$apply$8.apply(TestFramework.scala:202)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1$$anonfun$apply$8.apply(TestFramework.scala:202)
	at sbt.TestFramework$.sbt$TestFramework$$withContextLoader(TestFramework.scala:185)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1.apply(TestFramework.scala:202)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1.apply(TestFramework.scala:202)
	at sbt.TestFunction.apply(TestFramework.scala:207)
	at sbt.Tests$.sbt$Tests$$processRunnable$1(Tests.scala:239)
	at sbt.Tests$$anonfun$makeSerial$1.apply(Tests.scala:245)
	at sbt.Tests$$anonfun$makeSerial$1.apply(Tests.scala:245)
	at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
	at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
[error] Could not run test org.scalactic.anyvals.PosZFloatSpec: java.net.SocketException: Socket closed
java.lang.NullPointerException
	at sbt.JUnitXmlTestsListener$TestSuite$$anonfun$2.apply(JUnitXmlTestsListener.scala:78)
	at sbt.JUnitXmlTestsListener$TestSuite$$anonfun$2.apply(JUnitXmlTestsListener.scala:71)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
	at scala.collection.immutable.List.foreach(List.scala:318)
	at scala.collection.generic.TraversableForwarder$class.foreach(TraversableForwarder.scala:32)
	at scala.collection.mutable.ListBuffer.foreach(ListBuffer.scala:45)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:244)
	at scala.collection.AbstractTraversable.map(Traversable.scala:105)
	at sbt.JUnitXmlTestsListener$TestSuite.stop(JUnitXmlTestsListener.scala:71)
	at sbt.JUnitXmlTestsListener.writeSuite(JUnitXmlTestsListener.scala:180)
	at sbt.JUnitXmlTestsListener.endGroup(JUnitXmlTestsListener.scala:161)
	at sbt.TestRunner$$anonfun$run$4.apply(TestFramework.scala:90)
	at sbt.TestRunner$$anonfun$run$4.apply(TestFramework.scala:90)
	at sbt.TestFramework$$anonfun$safeForeach$1.apply(TestFramework.scala:107)
	at sbt.TestFramework$$anonfun$safeForeach$1.apply(TestFramework.scala:107)
	at scala.collection.immutable.List.foreach(List.scala:318)
	at sbt.TestFramework$.safeForeach(TestFramework.scala:107)
	at sbt.TestRunner.safeListenersCall(TestFramework.scala:96)
	at sbt.TestRunner.run(TestFramework.scala:90)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1$$anonfun$apply$8.apply(TestFramework.scala:202)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1$$anonfun$apply$8.apply(TestFramework.scala:202)
	at sbt.TestFramework$.sbt$TestFramework$$withContextLoader(TestFramework.scala:185)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1.apply(TestFramework.scala:202)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1.apply(TestFramework.scala:202)
	at sbt.TestFunction.apply(TestFramework.scala:207)
	at sbt.Tests$.sbt$Tests$$processRunnable$1(Tests.scala:239)
	at sbt.Tests$$anonfun$makeSerial$1.apply(Tests.scala:245)
	at sbt.Tests$$anonfun$makeSerial$1.apply(Tests.scala:245)
	at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
	at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
[error] java.lang.NullPointerException
java.net.SocketException: Socket closed
	at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:118)
	at java.net.SocketOutputStream.write(SocketOutputStream.java:155)
	at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
	at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
	at java.io.DataOutputStream.flush(DataOutputStream.java:123)
	at java.io.FilterOutputStream.close(FilterOutputStream.java:158)
	at scala.scalanative.testinterface.ComRunner.close(ComRunner.scala:107)
	at scala.scalanative.testinterface.ComRunner.send(ComRunner.scala:63)
	at scala.scalanative.testinterface.ScalaNativeRunner.send(ScalaNativeRunner.scala:33)
	at scala.scalanative.testinterface.ScalaNativeRunner.done(ScalaNativeRunner.scala:51)
	at sbt.Defaults$$anonfun$allTestGroupsTask$1$$anonfun$16.apply(Defaults.scala:625)
	at sbt.Defaults$$anonfun$allTestGroupsTask$1$$anonfun$16.apply(Defaults.scala:623)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
	at scala.collection.immutable.Map$Map2.foreach(Map.scala:130)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:244)
	at scala.collection.AbstractTraversable.map(Traversable.scala:105)
	at sbt.Defaults$$anonfun$allTestGroupsTask$1.apply(Defaults.scala:623)
	at sbt.Defaults$$anonfun$allTestGroupsTask$1.apply(Defaults.scala:621)
	at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
	at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

However, in 0.3.8, we no longer can skip the 2 passes using nativeOptimizerDriver like in 0.3.6, and it the same crash problem seems still persist in 0.3.8 after we commented out the nativeOptimizerDriver to get it compiled and linked.

We wonder if the crash problem will be addressed in 0.3.9 or is there anyway we can skip the 2 passes in 0.3.8?

Thanks.

Chee Seng

**Comment by ekrich** [comment]
@cheeseng It looks like as part of the Build API changes in `0.3.8` you may be able to change the code in your build slightly to do the same thing. The driver is now `scalanative.optimizer.Driver`. Referring to the following should help. https://github.com/scala-native/scala-native/blob/v0.3.8/tools/src/main/scala/scala/scalanative/optimizer/Driver.scala

**Comment by cheeseng** [comment]
@ekrich Actually it seems like nativeOptimizerDriver from ScalaNativePluginInternal is no longer accessible from the build, I tried both "debug" and "release" mode but unfortunately both mode also encountered the problem, probably because both of the mode contains DeadCodeElimination, I think.

Do you mind to enlighten us is there a way to bypass DeadCodeElimination pass in 0.3.8?

Thanks for your reply!

**Comment by ekrich** [comment]
I thought things got moved around a bit and didn't realize the `key` got removed from the internal plugin API. I'm afraid this is over my head at this point. Sorry I can't help more.

**Comment by densh** [comment]
@cheeseng Thanks for the bug report, this should not happen!

**Comment by LeeTibbert** [comment]
@cheeseng 

Thank you for the bug report and, especially, the URL to the repository where you were/are
encountering the problem.

@lolgab directed my attention to this open Issue.  I cloned the scalatest repository and,
after stumbling over an unfortunate scala/jdk11 interaction, built scalatest.


#### Short story

ScalaTest appears to run on Scala Native 0.3.8 just fine, for some values of fine.
If you do not experience the same on your system, I standby to help
Getting ScalaTest for Scala Native shipping at an early date would be a big success
all around.

#### Details

From reading the original report, I think that SN build passes are entirely innocent.
I suspect that you got captured by a known quirk of the SN unit-tests in SN 0.3.8.
That version of unit-tests did not like running too many tests at once.

 SN 0.4.0-M2 has a fix for that issue. I do not know if is a total fix, but it at lest raises the limit.
I hope to someday change unit-tests to behave better than "Socket closed" when they
encounter fatal (to it) errors.  The message you saw in the base not is neither kind
nor informative to those who have lead a good, clean life.

Suspecting a SN unit-test issue, I ran 15 tests Suites described at
 (URL) [https://github.com/scalatest/scalatest] each in their own sbt invocation
(with Java 8).  My thanks to the author/s of that page.  Each test Suite appeared
to run to successful completion.

Not content with success, I then removed the gentests directory, as directed,
and ran all the tests in the same sbt run: 
`$ sbt test // with some JAVA_HOME for JDK 8 stuff`

That also appeared to mostly work; that is, no obvious SN socket errors as in the 
base report.  Here is where I need your expertise. Three tests outright failed:
```
[error] Failed: Total 18020, Failed 3, Errors 0, Passed 18017, Ignored 50, Canceled 11, Pending 42
[error] Failed tests:
[error] 	org.scalatest.concurrent.ConductorFixtureDeprecatedSuite
[error] 	org.scalatest.concurrent.ConductorMethodsDeprecatedSuite
[error] 	org.scalatest.concurrent.WaitersSpec
```
SN unit-tests sometimes tells alternate truths about which tests failed so I examined the
log file.  It looks like these failed with timeouts. That is understandable because 
SN is currently single threaded.    I think the `[error] Failed;` on the first line is just
SBT unhappy with having been given output by one or more of the Suites.

Are these Suites important/critical for ScalaTest? I did not look at the contends
of the Suites.

When I grepped the log file for `,  failed` I saw lots of lines with what I think are 
"expected errors".

What do you experience in your environment?  Do problems remain?

NB:   I tried to replicate the original report as close as possible.   I tried none of
         0.3.9, 0.4.0-M2, or 0.4.0-SNAPSHOT (> M2).  I thought I would leave the 
         enjoyment of figuring out the tangle of unpublished dependencies for another
         day.

**Comment by LeeTibbert** [comment]
@cheeseng , @lolgab, @densh 

I deleted my post of today because I discovered studying the log file that the tests were
 being run under JVM, not SN.  Sorry for taking your time.

I will take a look at this again and see if I can get the tests running using SN.

Some days you get the bear; some days the bear gets you.

**Comment by LeeTibbert** [comment]
I _still_ think this is a unit-tests problem but need to gather evidence (and be sure I am
actually running SN).

**Comment by cheeseng** [comment]
@LeeTibbert @lolgab @densh To build and run SN tests, you should do (search for 'native' in README): 

```
> sbt ++2.11.12 scalatestAppNative/clean
> sbt ++2.11.12 scalatestTestNative/test:compile
> sbt ++2.11.12 scalatestTestNative/test
```

If you have enough system memory, the above should build with SN and the tests should run, you'll notice some failing tests (which I had logged with different tickets).  The problem reported for this ticket, if that if you try to change the followings to 0.3.8: 

https://github.com/scalatest/scalatest/blob/3.1.x/project/scalatest.scala#L1155
https://github.com/scalatest/scalatest/blob/3.1.x/project/scalatest.scala#L1451
https://github.com/scalatest/scalatest/blob/3.1.x/project/plugins.sbt#L13

It'll breaks the build, it seems like the following (as https://github.com/cheeseng/scalatest/blob/experiment-scala-native-0.3.8/project/scalatest.scala#L679 in the report description above): 

```
nativeOptimizerDriver in NativeTest := {
        val orig = tools.OptimizerDriver((nativeConfig in NativeTest).value)
        orig.withPasses(orig.passes.filterNot(p => p == pass.DeadBlockElimination || p == pass.GlobalBoxingElimination))
      }
```

has been taken out but without an alternative way to achieve the same thing in 0.3.8.

Hope this clarify the build and problem.

And thanks for looking into this!

**Comment by LeeTibbert** [comment]
* @cheeseng Thank you for the instructions. Now that you point out that they exist, I was
   able to find them, bold as day, in the Fine Manual.  Duh! 

* I am now sure that I am exercising Scala Native.

* Again, I stand by to help get ScalaTest publish for current Scala Native versions.

#### Solution to base problem:

The reported problem, and its siblings in various other versions are almost certainly
`Out-of-Memory` failures, poorly reported by Scala Native.  The ScalaTest README
suggests increasing the memory when running the tests. I followed those instructions
and used a bigger value.  I was going for success, so I went big and did not try
to titrate the value. I also dropped some no longer relevant JVM Garbage collection.
```
-server 
-Xms512M 
-Xmx15G 
-Xss10M
-XX:+CMSClassUnloadingEnabled 
```

With this change, the experimental branch in the base note works with Scala Native 0.3.8.
That is, the TestRunner completes. I believe both test runs detected some failures
in the test cases (1000 or so for he scalatestNativeTest).  Those are another problem
(see below).

#### Notes

* Somewhere I got the idea that ScalaTest had a dependency on ScalaCheck and that 
  this would be a problem for current Scala Native versions.  @lolgab I encountered
  no such dependency. I do not know if my thought was wrong or if the dependency has
  been removed.  Seems to be no blocker here.

* I do not know what the SN plugin phase manipulation in the ScalaTest build before the 
  base note was trying to accomplish.  The base note had that code commented out,
  so I deleted it in the attempts below.  Seems not to be required not, but not speaking,
  not knowing.

* I cloned the ScalaTest 3.1.x branch and ran a series of experiments:

  * It builds and runs nicely with ScalaNative 0.4.0-M2 (and the above .jvmopt memory
    change).  Both TestRunners completed. scalacticTestNative/test reported 4 failing
    unit tests and scalatestTestNative/test reported ZERO.  

    I did not examine the failing unit-tests, a problem for day after this issue has settled.
    Guess I get a chance to learn what Scalactic does.  

  * it builds and runs with ScalaNative 0.3.8.  Both TestRunners completed.  There
    were about a thousand failing unit-tests out of 8K or total.  

  * it builds and mostly runs on ScalaNative 0.3.9 (current tip). I say mostly because
    both compiles and the scalatestTestNative/test passes completed as expected
    but the scalacticTestNative/test failed with `at Heap_exitWithOutOfMemory`
    after consuming approx. 15G of physical memory and 10+G of swap space.
   
    This result surprises me since there were not changes from 0.3.8 to 0.3.9
    should give it.  Famous dying cry of the devo....

  * I believe it has link issues ScalaNative 0.4.0-SNAPSHOT (2019-07-28).
    If I can replicate what looked like a SN name mangling or inlining bug I 
    will report it in a new issue.

**Comment by LeeTibbert** [comment]
Thoughts for the longer run:

Not everything can be done at once, but when the cycles become available, looking at the
entire SBT/ScalaTest/SN_TestRunner stack will be worthwhile.  This looks to me
like an example where "programming in the small" does not scale well to "programming in the large".  What worked well as bootstrap fails in industrial use; part of the Software Life Cycle.

The key issue is "What can be done to reduce memory usage when LARGE numbers, 
meaning thousands or tens of thousands, of Suites and test cases are run"?

Currently SN runs all the test Suites in one process.  This assumes that each 
Suite  is well behaved in its memory usage.  Leaks & plain abuse can accumulate
over the entire test run.  Can the remote runner be (economically) articulated to run
each Suite in a sub or, better yet, independent, process?

What happens,in detail, when ScalaTest is used as the test Framework in the remote process?
I've certainly hung my machine, requiring a power cycle, using 15G of RAM and 20G of swap.

So, not today's problem, but one that will certainly cycle around and re-revisit.

**Comment by LeeTibbert** [comment]
For the sakes of completeness & reproducibility, I should say that the results above were
all obtained in DEBUG mode, on a current Linux Ubuntu system.

**Comment by cheeseng** [comment]
@LeeTibbert Thanks for the great info and summary, I tried to use 0.4.0-M2 and it works nicely, I have submitted a PR to 3.1.x branch here: 

https://github.com/scalatest/scalatest/pull/1673

I think the original problem of upgrading to 0.3.8 is now no longer relevant, closing this.

Thanks again for your great help!

**Comment by muxanick** [comment]
FYI, I am seeing this error on latest github version, after running `tests/test` a few times (approx. 20 times)

**Comment by LeeTibbert** [comment]
@cheeseng  Great to hear that ScalaTest will have support for Scala Native. I have a part of
ScalaNative developement in mind which has been blocked by lack of ScalaTest. 
Thank you for investing in Scala Native.  

Let me know if I can be helpful in helping get ScalaTest with SN support out the door.

@muxanick   

Sorry our are experiencing a problem. To keep things from becoming tangled, can
you create a new issue?  If you do, I will read it.

There is a _very_ good chance the response will be the essence of the solution in 
this case, see below.  However, I want to look at your primary report and its data
before reaching a conclusion.  It would be nice to keep your Windows SN work rollling.
```
Solution to base problem:

The reported problem, and its siblings in various other versions are almost certainly
Out-of-Memory failures, poorly reported by Scala Native. The ScalaTest README
suggests increasing the memory when running the tests. I followed those instructions
and used a bigger value. I was going for success, so I went big and did not try
to titrate the value. I also dropped some no longer relevant JVM Garbage collection.

-server 
-Xms512M 
-Xmx15G 
-Xss10M
-XX:+CMSClassUnloadingEnabled 

With this change, the experimental branch in the base note works with Scala Native 0.3.8.
That is, the TestRunner completes. I believe both test runs detected some failures
in the test cases (1000 or so for he scalatestNativeTest). Those are another problem
(see below).
```
